pkg/tls/config.go

```diff
@@ -0,0 +1,214 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// TODO
+type WatchedConfig struct {
+	log         logrus.FieldLogger
+	caFiles     []string
+	certFile    string
+	privkeyFile string
+	stop        chan struct{}
+	mu          lock.Mutex
+	caCertPool  *x509.CertPool
+	keypair     *tls.Certificate
+}
+
+var (
+	CertWithoutPrivkeyErr = errors.New("certificate and private key are both required, but only the certificate was provided")
+	PrivkeyWithoutCertErr = errors.New("certificate and private key are both required, but only the private key was provided")
+)
+
+// TODO
+func NewWatchedConfig(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (*WatchedConfig, error) {
+	if certFile != "" && privkeyFile == "" {
+		return nil, CertWithoutPrivkeyErr
+	}
+	if certFile == "" && privkeyFile != "" {
+		return nil, PrivkeyWithoutCertErr
+	}
+
+	cfg := &WatchedConfig{
+		log:         log,
+		caFiles:     caFiles,
+		certFile:    certFile,
+		privkeyFile: privkeyFile,
+		stop:        make(chan struct{}),
+	}
+
+	// load the files for the first time before starting the watcher and update
+	// goroutine.
+	if err := cfg.update(); err != nil {
+		return nil, err
+	}
+	if err := cfg.start(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// start initialize the WatchedConfig files watcher and update goroutine.
+func (cfg *WatchedConfig) start() error {
+	// file watcher setup.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("Failed to create fsnotify watcher: %s", err)
+	}
+	for _, path := range cfg.caFiles {
+		if err := watcher.Add(path); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", path, err)
+		}
+	}
+	if cfg.certFile != "" {
+		if err := watcher.Add(cfg.certFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.certFile, err)
+		}
+	}
+	if cfg.privkeyFile != "" {
+		if err := watcher.Add(cfg.privkeyFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.privkeyFile, err)
+		}
+	}
+
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case event := <-watcher.Events:
+				cfg.log.Debugf("Received fsnotify event: %+v", event)
+				switch event.Op {
+				case fsnotify.Create, fsnotify.Write, fsnotify.Chmod, fsnotify.Remove, fsnotify.Rename:
+					if err := cfg.update(); err != nil {
```

Comment:
I wonder if it would make sense to decompose `cfg.update()` into more specific functions. When receiving a `fsnotify.Event`, every watched file is reloaded with the current implementation. Using `event.Name`, I think we could reload only the file that should be reloaded.

Reply:
I elected for simplicity to start with in order to focus on the overall design, but your suggestion make sense.